src/eduid_webapp/email/views.py

```diff
@@ -0,0 +1,295 @@
+            'error': {'form': 'mail_duplicated'}
+        }
+    try:
+        save_dashboard_user(user, dbattr_name='dashboard_userdb')
+    except UserOutOfSync:
+        current_app.logger.debug('Couldnt save email {!r} for user {!r}, '
+                            'data out of sync'.format(email, user))
+        return {
+            '_status': 'error',
+            'error': {'form': 'out_of_sync'}
+        }
+    current_app.logger.info('Saved unconfirmed email {!r} '
+                            'for user {!r}'.format(email, user))
+
+    send_verification_code(email, user)
+
+    emails = {'emails': user.mail_addresses.to_list_of_dicts()}
+    return EmailListPayload().dump(emails).data
+
+
+@email_views.route('/primary', methods=['POST'])
+@UnmarshalWith(SimpleEmailSchema)
+@MarshalWith(EmailResponseSchema)
+@require_dashboard_user
+def post_primary(user, email, csrf_token):
+    """
+    """
+    if session.get_csrf_token() != csrf_token:
+        abort(400)
+    current_app.logger.debug('Trying to save email address {!r} as primary '
+                             'for user {!r}'.format(email, user))
+    try:
+        mail = user.mail_addresses.find(email)
+    except IndexError:
+        current_app.logger.debug('Couldnt save email {!r} as primary for user'
+                                 ' {!r}, data out of sync'.format(email, user))
+        return {
+            '_status': 'error',
+            'error': {'form': 'out_of_sync'}
+        }
+
+    if not mail.is_verified:
+        current_app.logger.debug('Couldnt save email {!r} as primary for user'
+                                ' {!r}, email unconfirmed'.format(email, user))
+        return {
+            '_status': 'error',
+            'error': {'form': 'emails.unconfirmed_address_not_primary'}
+        }
+
+    user.mail_addresses.primary = mail.email
+    try:
+        save_dashboard_user(user, dbattr_name='dashboard_userdb')
+    except UserOutOfSync:
+        current_app.logger.debug('Couldnt save email {!r} as primary for user'
+                                 ' {!r}, data out of sync'.format(email, user))
+        return {
+            '_status': 'error',
+            'error': {'form': 'out_of_sync'}
+        }
+    current_app.logger.info('Email address {!r} made primary '
+                            'for user {!r}'.format(email, user))
+    emails = {'emails': user.mail_addresses.to_list_of_dicts()}
+    return EmailListPayload().dump(emails).data
+
+
+@email_views.route('/verify', methods=['POST'])
+@UnmarshalWith(VerificationCodeSchema)
+@MarshalWith(EmailResponseSchema)
+@require_dashboard_user
+def verify(user, code, email, csrf_token):
+    """
+    """
+    if session.get_csrf_token() != csrf_token:
+        abort(400)
+    current_app.logger.debug('Trying to save email address {!r} as verified '
+                             'for user {!r}'.format(email, user))
+    db = current_app.verifications_db
+    state = db.get_state_by_eppn_and_code(user.eppn, code)
```

Comment:
I think this should be get_state_by_eppn_and_email. That way we have to make sure the database only contains one verification entry per user and mail address. There is no need to keep the verifications as we did in the verifications collection in the old Dashboard app.

Also even if we can most likely promise that the verifications codes are unique this could be a security problem with shorter codes.